noir-projects/noir-contracts/contracts/protocol/tagging_contract/src/util.nr

```diff
@@ -0,0 +1,102 @@
+fn get_next_tag(
+    sender: AztecAddress,
+    recipient: AztecAddress,
+    hidden_sender: bool,
+    context: &mut PrivateContext,
+) -> Field {
+    if !handshake_exists(sender, recipient, hidden_sender) {
+        // The handshake doesn't exist, so we need to call the handshake function of the Handshaker contract.
+        context.call_private_function(
+            HANDSHAKER_CONTRACT_ADDRESS,
+            FunctionSelector::from_signature("handshake((Field,bool))"),
+            [sender.to_field(), recipient.to_field(), hidden_sender as Field],
+        );
+    }
+
+    let shared_tagging_public_key = get_shared_tagging_public_key(sender, recipient, hidden_sender);
+
+    let sender_npk_m = get_public_keys(sender).npk_m;
+    let sender_nsk_app = context.request_nsk_app(sender_npk_m.hash());
+
+    // We call the is_first_time_tagging oracle function to determine if this is the first time tagging
+    // for [sender, recipient, hidden_sender] inputs. The function computes the tag nullifier with index
+    // 0 and checks if it exists in the nullifier tree.
+    let first_time_tagging: bool = is_first_time_tagging(sender, recipient, hidden_sender);
+
+    let (unsiloed_nullifier_preimage, app_to_silo_with, directional_app_tagging_secret, index) = if first_time_tagging {
+        // Sort addresses to reliably get bi-directional handshake commitment. We do that in an unconstrained function
+        // and then we constrain the result to optimize constraints.
+        let (address_0, address_1) = sort_addresses(sender, recipient);
+
+        let handshake_commitment_preimage = [
+            "AZTEC_NR::HANDSHAKE_SEPARATOR",
+            shared_tagging_public_key.x,
+            shared_tagging_public_key.y,
+            address_0,
+            address_1,
+        ];
```

Comment:
Good thing is that both the `handshake_commitment_preimage` here and the `previous_tag_nullifier_preimage` in the else branch have the same length so we can easily just return it from the if-else.